federatedscope/contrib/optimizer/lomo.py

```diff
@@ -0,0 +1,253 @@
+'''
+This implementation is adapted from https://github.com/OpenLMLab/LOMO
+(MIT License)
+
+Copyright (c) 2023 OpenLMLab
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+'''
+
+try:
+    import torch
+    from torch.optim import Optimizer
+except ImportError:
+    torch = None
+
+from federatedscope.register import register_optimizer
+
+
+class LOMO(Optimizer):
+    """
+    an optimizer for LOMOTrainer
+    """
+    def __init__(self,
+                 model,
+                 lr=1e-3,
+                 clip_grad_norm=None,
+                 clip_grad_value=None):
+        if torch is None:
+            raise ImportError(
+                'Currently, LOMO optimizer is only implemented with `pytorch`')
+        self.model = model
+        self.lr = lr
+        self.local_rank = 0
+        self.world_size = 1
+        self.clip_grad_norm = clip_grad_norm
+        self.clip_grad_value = clip_grad_value
+
+        # for grad norm
+        if self.clip_grad_norm is not None and self.clip_grad_norm <= 0:
+            raise ValueError("clip_grad_norm should be positive,"
+                             f" got {self.clip_grad_norm}.")
+        self.gather_norm = False
+        self.grad_norms = []
+        self.clip_coef = None
+
+        # check if zero3 is enabled
+        p0 = list(self.model.parameters())[0]
+        if hasattr(p0, 'ds_tensor'):  # zero3 is enabled
```

Comment:
perhaps add some reference to make the checking reasonable, e.g., [stage3_code](https://deepspeed.readthedocs.io/en/stable/_modules/deepspeed/runtime/zero/stage3.html)

Reply:
Many thanks. This reference has been added in the latest commit.